File: PE-common/src/main/java/nl/adaptivity/process/processModel/ActivityBase.java
/*
 * Copyright (c) 2016.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with Foobar.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.process.processModel;

import net.devrieze.util.StringUtil;
import nl.adaptivity.process.ProcessConsts.Engine;
import nl.adaptivity.process.util.Identifiable;
import nl.adaptivity.process.util.Identifier;
import nl.adaptivity.util.xml.SimpleXmlDeserializable;
import nl.adaptivity.util.xml.XmlUtil;
import nl.adaptivity.xml.XmlException;
import nl.adaptivity.xml.XmlReader;
import nl.adaptivity.xml.XmlWriter;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.xml.namespace.QName;

import java.util.Collection;
import java.util.Collections;


/**
 * Base class for activity implementations
 * Created by pdvrieze on 23/11/15.
 */
public abstract class ActivityBase<T extends ProcessNode<T, M>, M extends ProcessModelBase<T, M>> extends ProcessNodeBase<T, M> implements Activity<T, M>, SimpleXmlDeserializable {

  @Nullable private XmlMessage mMessage;

  @Nullable private String mName;

  public ActivityBase(final Activity<?, ?> orig) {
    super(orig);
    setMessage(orig.getMessage());
    setName(orig.getName());
  }

  // Object Initialization
  public ActivityBase(@Nullable final M ownerModel) {
    super(ownerModel);
  }
// Object Initialization end

  @Override
  public final <R> R visit(@NotNull final Visitor<R> visitor) {
    return visitor.visitActivity(this);
  }

  @Override
  public boolean deserializeChild(@NotNull final XmlReader in) throws XmlException {
    if (Engine.NAMESPACE.equals(in.getNamespaceUri())) {
      switch (in.getLocalName().toString()) {
        case XmlDefineType.ELEMENTLOCALNAME:
          getDefines().add(XmlDefineType.deserialize(in));
          return true;
        case XmlResultType.ELEMENTLOCALNAME:
          getResults().add(XmlResultType.deserialize(in));
          return true;
        case Condition.ELEMENTLOCALNAME:
          deserializeCondition(in);
          return true;
        case XmlMessage.ELEMENTLOCALNAME:
          setMessage(XmlMessage.deserialize(in));
          return true;
      }
    }
    return false;
  }

  protected abstract void deserializeCondition(final XmlReader in) throws XmlException;

  @Override
  public boolean deserializeChildText(final CharSequence elementText) {
    return false;
  }

  @Override
  public final void serialize(@NotNull final XmlWriter out) throws XmlException {
    XmlUtil.writeStartElement(out, ELEMENTNAME);
    serializeAttributes(out);
    serializeChildren(out);
    XmlUtil.writeEndElement(out, ELEMENTNAME);
  }

  @Override
  protected final void serializeAttributes(@NotNull final XmlWriter out) throws XmlException {
    super.serializeAttributes(out);
    Identifiable predecessor = getPredecessor();
    if (predecessor!=null) {
      XmlUtil.writeAttribute(out, ATTR_PREDECESSOR, predecessor.getId());
    }
    XmlUtil.writeAttribute(out, "name", getName());
  }

  protected final void serializeChildren(final XmlWriter out) throws XmlException {
    super.serializeChildren(out);
    serializeCondition(out);

    {
      final XmlMessage m = getMessage();
      if (m != null) { m.serialize(out); }
    }
  }

  protected abstract void serializeCondition(final XmlWriter out) throws XmlException;

  @Override
  public final boolean deserializeAttribute(final CharSequence attributeNamespace, @NotNull final CharSequence attributeLocalName, final CharSequence attributeValue) {
    switch (attributeLocalName.toString()) {
      case ATTR_PREDECESSOR:
        setPredecessor(new Identifier(attributeValue.toString()));
        return true;
      case "name":
        setName(StringUtil.toString(attributeValue));
        return true;
    }
    return super.deserializeAttribute(attributeNamespace, attributeLocalName, attributeValue);
  }

  @Override
  public final void setDefines(@Nullable final Collection<? extends IXmlDefineType> exports) {
    super.setDefines(exports);
  }

  @Override
  public final void setResults(@Nullable final Collection<? extends IXmlResultType> imports) { super.setResults(imports); }

  /* (non-Javadoc)
       * @see nl.adaptivity.process.processModel.IActivity#getName()
       */
  @Override
  public final String getName() {
    return mName;
  }

  /* (non-Javadoc)
     * @see nl.adaptivity.process.processModel.IActivity#setName(java.lang.String)
     */
  @Override
  public final void setName(final String name) {
    mName = name;
  }

  /* (non-Javadoc)
       * @see nl.adaptivity.process.processModel.IActivity#getPredecessor()
       */
  @Nullable
  @Override
  public final Identifiable getPredecessor() {
    final Collection<? extends Identifiable> ps = getPredecessors();
    if ((ps == null) || (ps.size() != 1)) {
      return null;
    }
    return ps.iterator().next();
  }

  @Override
  public final void setPredecessor(final Identifiable predecessor) {
    setPredecessors(Collections.singleton(predecessor));
  }

  @Override
  public final XmlMessage getMessage() {
    return mMessage;
  }

  @Override
  public final void setMessage(final IXmlMessage message) {
    setMessage(XmlMessage.get(message));
  }

  public final void setMessage(final XmlMessage message) {
    mMessage = XmlMessage.get(message);
  }

  @NotNull
  @Override
  public final QName getElementName() {
    return ELEMENTNAME;
  }

  @Override
  public boolean equals(final Object o) {
    if (this == o) { return true; }
    if (o == null || getClass() != o.getClass()) { return false; }
    if (!super.equals(o)) { return false; }

    ActivityBase<?, ?> that = (ActivityBase<?, ?>) o;

    if (mMessage != null ? !mMessage.equals(that.mMessage) : that.mMessage != null) { return false; }
    final String condition = getCondition();
    if (condition != null ? ! condition.equals(that.getCondition()) : that.getCondition() != null) { return false; }
    return mName != null ? mName.equals(that.mName) : that.mName == null;

  }

  @Override
  public int hashCode() {
    int result = super.hashCode();
    result = 31 * result + (mMessage != null ? mMessage.hashCode() : 0);
    final String condition = getCondition();
    result = 31 * result + (condition != null ? condition.hashCode() : 0);
    result = 31 * result + (mName != null ? mName.hashCode() : 0);
    return result;
  }

  // Property acccessors end
}

File: PE-common/src/main/kotlin/nl/adaptivity/process/processModel/ActivityBase.kt
/*
 * Copyright (c) 2016.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.process.processModel

import net.devrieze.util.collection.replaceBy
import nl.adaptivity.process.ProcessConsts.Engine
import nl.adaptivity.process.processModel.engine.XmlCondition
import nl.adaptivity.process.util.Identifiable
import nl.adaptivity.process.util.Identified
import nl.adaptivity.process.util.Identifier
import nl.adaptivity.util.xml.SimpleXmlDeserializable
import nl.adaptivity.xml.*
import javax.xml.namespace.QName


/**
 * Base class for activity implementations
 * Created by pdvrieze on 23/11/15.
 */
abstract class ActivityBase<NodeT : ProcessNode<NodeT, ModelT>, ModelT : ProcessModel<NodeT, ModelT>?> : ProcessNodeBase<NodeT, ModelT>, Activity<NodeT, ModelT> {

  abstract class Builder<NodeT : ProcessNode<NodeT, ModelT>, ModelT: ProcessModel<NodeT, ModelT>?> : ProcessNodeBase.Builder<NodeT,ModelT>, Activity.Builder<NodeT,ModelT>, SimpleXmlDeserializable {

    override var message: IXmlMessage?
    override var name: String?
    override var condition: String?
    override val idBase:String
        get() = "ac"

    override var childId: String? = null

    constructor(): this(id = null)

    constructor(id: String? = null,
                predecessor: Identified? = null,
                successor: Identified? = null,
                label: String? = null,
                defines: Collection<IXmlDefineType> = emptyList(),
                results: Collection<IXmlResultType> = emptyList(),
                message: XmlMessage? = null,
                condition: String? = null,
                name: String? = null,
                x: Double = Double.NaN,
                y: Double = Double.NaN) : super(id, listOfNotNull(predecessor), listOfNotNull(successor), label, defines, results, x, y) {
      this.message = message
      this.name = name
      this.condition = condition
    }

    constructor(node: Activity<*, *>) : super(node) {
      this.message = XmlMessage.get(node.message)
      this.name = node.name
      this.condition = node.condition
      this.childId = node.childModel?.id
    }

    override val elementName: QName get() = Activity.ELEMENTNAME

    @Throws(XmlException::class)
    override fun deserializeChild(reader: XmlReader): Boolean {
      if (Engine.NAMESPACE == reader.namespaceUri) {
        when (reader.localName.toString()) {
          XmlDefineType.ELEMENTLOCALNAME -> (defines as MutableList).add(XmlDefineType.deserialize(reader))

          XmlResultType.ELEMENTLOCALNAME -> (results as MutableList).add(XmlResultType.deserialize(reader))

          Condition.ELEMENTLOCALNAME -> condition = XmlCondition.deserialize(reader).condition

          XmlMessage.ELEMENTLOCALNAME -> message=XmlMessage.deserialize(reader)

          else -> return false
        }
        return true
      }
      return false
    }

    override fun deserializeAttribute(attributeNamespace: CharSequence, attributeLocalName: CharSequence, attributeValue: CharSequence): Boolean {
      when (attributeLocalName.toString()) {
        ProcessNodeBase.ATTR_PREDECESSOR -> predecessors.replaceBy(Identifier(attributeValue.toString()))
        "name" -> name = attributeValue.toString()
        ATTR_CHILDID -> childId = attributeValue.toString()
        else -> return super<ProcessNodeBase.Builder>.deserializeAttribute(attributeNamespace, attributeLocalName, attributeValue)
      }
      return true
    }

    override fun deserializeChildText(elementText: CharSequence): Boolean {
      return false
    }

    override fun toString(): String {
      return "${super.toString().dropLast(1)}, message=$message, name=$name, condition=$condition)"
    }


  }

  abstract class ChildModelBuilder<NodeT : ProcessNode<NodeT, ModelT>, ModelT: ProcessModel<NodeT, ModelT>?>(
      id: String? = null,
      override var childId: String? = null,
      nodes: Collection<ProcessNode.IBuilder<NodeT, ModelT>> = emptyList(),
      predecessors: Collection<Identified> = emptyList(),
      successors: Collection<Identified> = emptyList(),
      label: String? = null,
      imports: Collection<IXmlResultType> = emptyList(),
      defines: Collection<IXmlDefineType> = emptyList(),
      exports: Collection<IXmlDefineType> = emptyList(),
      results: Collection<IXmlResultType> = emptyList(),
      x: Double = Double.NaN,
      y: Double = Double.NaN) : ProcessNodeBase.Builder<NodeT, ModelT>(id, predecessors, successors, label, defines, results, x, y), Activity.ChildModelBuilder<NodeT,ModelT> {
    override val nodes: MutableSet<ProcessNode.IBuilder<NodeT, ModelT>> = nodes.toMutableSet()
    override val imports: MutableList<IXmlResultType> = imports.toMutableList()
    override val exports: MutableList<IXmlDefineType> = exports.toMutableList()
    override var condition: String? = null

    override val idBase:String get() = "child"

    override val elementName: QName get() = ChildProcessModel.ELEMENTNAME

    override fun deserializeAttribute(attributeNamespace: CharSequence, attributeLocalName: CharSequence, attributeValue: CharSequence)=false

  }


  private var _message: XmlMessage? = null

  private var _name: String? = null

  override val childModel: ChildProcessModel<NodeT, ModelT>?

  override var name:String?
    get() = _name
    set(value) { _name = value}

  override var predecessor: Identifiable?
    get() = if (predecessors.isEmpty()) null else predecessors.single()
    set(value) {
      setPredecessors(listOfNotNull(value?.identifier))
    }

  override var message: IXmlMessage?
    get() = _message
    set(value) {
      _message = XmlMessage.get(value)
    }

  fun setMessage(message: XmlMessage?) {
    _message = message
  }

  constructor(builder: Activity.Builder<*, *>, buildHelper: ProcessModel.BuildHelper<NodeT, ModelT>) : super(builder, buildHelper) {
    if(builder.message!=null && builder.childId!=null) throw IllegalProcessModelException("Activities can not have child models as well as messages")
    this._message = XmlMessage.get(builder.message)
    this._name = builder.name
    this.childModel = builder.childId?.let{ buildHelper.childModel(it) }
  }

  constructor(builder: Activity.ChildModelBuilder<*, *>, buildHelper: ProcessModel.BuildHelper<NodeT, ModelT>) : super(builder, buildHelper) {
    this._message = null
    this._name = null
    this.childModel = buildHelper.childModel(builder.childId!!)
  }
  // Object Initialization end

  override abstract fun builder(): Builder<NodeT, ModelT>

  override fun <R> visit(visitor: ProcessNode.Visitor<R>): R {
    return visitor.visitActivity(this)
  }

  @Throws(XmlException::class)
  override fun serialize(out: XmlWriter) {
    out.smartStartTag(Activity.ELEMENTNAME)
    serializeAttributes(out)
    serializeChildren(out)
    out.endTag(Activity.ELEMENTNAME)
  }

  @Throws(XmlException::class)
  override fun serializeAttributes(out: XmlWriter) {
    super.serializeAttributes(out)
    out.writeAttribute(ProcessNodeBase.ATTR_PREDECESSOR, predecessor?.id)
    out.writeAttribute("childId", childModel?.id)
    out.writeAttribute("name", name)
  }

  @Throws(XmlException::class)
  override fun serializeChildren(out: XmlWriter) {
    super.serializeChildren(out)
    serializeCondition(out)

    _message?.serialize(out)
  }

  @Throws(XmlException::class)
  protected abstract fun serializeCondition(out: XmlWriter)

  /* Override to make public */
  override fun setDefines(defines: Collection<IXmlDefineType>) = super.setDefines(defines)

  /* Override to make public */
  override fun setResults(results: Collection<IXmlResultType>) = super.setResults(results)

  override fun equals(other: Any?): Boolean {
    if (this === other) {
      return true
    }
    if (other == null || javaClass != other.javaClass) {
      return false
    }
    if (!super.equals(other)) {
      return false
    }

    val that = other as ActivityBase<*, *>?

    if (_message != that!!._message) {
      return false
    }
    val condition = condition
    if (condition != that.condition) {
      return false
    }
    return _name == that._name

  }

  override fun hashCode(): Int {
    var result = super.hashCode()
    result = 31 * result + (_message?.hashCode() ?: 0)
    result = 31 * result + (condition?.hashCode() ?: 0)
    result = 31 * result + (_name?.hashCode() ?: 0)
    return result
  }

  // Property acccessors end

  companion object {
    const val ATTR_CHILDID = "childId"
  }

}
